File: problems/Number of Islands II/numIslands2.cs
public class Solution {
    public IList<int> NumIslands2(int m, int n, int[][] positions) {
        bool[,] grid = new bool[m,n];
        IList<int> res = new List<int>(positions.Length);
        ISet<int> roots = new HashSet<int>();
        Unions dsu = new Unions(m*n);

        IList<int> removeRoots = new List<int>();

        for (int i = 0; i < positions.Length; i++)
        {
            var pos = positions[i];
            if (grid[pos[0], pos[1]])
            {
                res.Add(roots.Count);
                continue;
            }


            grid[pos[0], pos[1]] = true;
            var posLinear = n * pos[0] + pos[1];
            removeRoots.Clear();

            foreach (var dir in _directions)
            {
                int newI = pos[0] + dir.di;
                int newJ = pos[1] + dir.dj;
                int linear = newI * n + newJ;

                if (newI >= 0 && newI < m && newJ >= 0 && newJ < n && grid[newI, newJ])
                {
                    removeRoots.Add(dsu.Find(linear));
                }
            }

            foreach (var dir in _directions)
            {
                int newI = pos[0] + dir.di;
                int newJ = pos[1] + dir.dj;
                int linear = newI * n + newJ;

                if (newI >= 0 && newI < m && newJ >= 0 && newJ < n && grid[newI, newJ])
                {
                    dsu.Union(posLinear, linear);
                }
            }

            foreach (var removeRoot in removeRoots)
            {
                roots.Remove(removeRoot);
            }

            roots.Add(dsu.Find(posLinear));

            res.Add(roots.Count);
        }

        return res;
    }

    private static readonly (int di, int dj)[] _directions = {(0, 1), (0, -1), (1, 0), (-1, 0)};

    private class Unions
    {
        private readonly int[] _parents;
        private readonly int[] _ranks;

        public Unions(int n)
        {
            _parents = new int[n];
            _ranks = new int[n];
            for (int i = 0; i < n; i++)
            {
                _parents[i] = i;
            }
        }

        public int Find(int x)
        {
            if (x != _parents[x])
            {
                x = Find(_parents[x]);
            }
            return _parents[x];
        }

        public bool Union(int x, int y)
        {
            int px = Find(x);
            int py = Find(y);
            if (px == py)
            {
                return false;
            }
            if (_ranks[px] > _ranks[py])
            {
                _parents[py] = px;
                _ranks[px]++;
            }
            else
            {
                _parents[px] = py;
                _ranks[py]++;
            }
            return true;
        }
    }
}

File: problems/Number of Islands II/NumIslands2.java
public class Solution {
    public IList<int> NumIslands2(int rows, int cols, int[][] positions) {
        var n = positions.Length;
        var result = new int[n];
        var grid = new bool[rows, cols];
        var dsu = new Dsu(1 + rows * cols);

        for (var i = 0; rows * cols > i; ++i) {
            dsu.union(i, rows * cols);
        }

        var connectedComponentsCount = 0;

        for (var i = 0; n > i; ++i) {
            var row = positions[i][0];
            var col = positions[i][1];

            if (!grid[row, col]) {
                var connections = 0;

                grid[row, col] = true;
                dsu.setParent(cols * row + col);

                if (0 < row && grid[row - 1, col]) {
                    connections+=dsu.union(cols*(row-1)+col,cols*row+col)?1:0;
                }
                if (0 < col && grid[row, col - 1]) {
                    connections+=dsu.union(cols*row-1+col,cols*row+col)?1:0;
                }
                if (rows > 1 + row && grid[1 + row, col]) {
                    connections+=dsu.union(cols*(1+row)+col,cols*row+col)?1:0;
                }
                if (cols > 1 + col && grid[row, 1 + col]) {
                    connections+=dsu.union(cols*row+1+col,cols*row+col)?1:0;
                }

                connectedComponentsCount -= (connections - 1);
            }

            result[i] = connectedComponentsCount;
        }

        return result.ToList();
    }

    private class Dsu {
        private readonly int[] _parents;
        private readonly int[] _ranks;

        public Dsu(int n) {
            _parents = new int[n];
            _ranks = new int[n];

            for (var i = 0; n > i; ++i) {
                _parents[i] = i;
            }
        }

        public int find(int x) {
            if (_parents[x] != x) {
                _parents[x] = find(_parents[x]);
            }

            return _parents[x];
        }

        public void setParent(int x) {
            var px = find(x);

            _parents[x] = x;
            --_ranks[px];
        }

        public bool union(int x, int y) {
            var px = find(x);
            var py = find(y);

            if (px == py) {
                return false;
            }

            if (_ranks[px] > _ranks[py]) {
                _parents[py] = px;
                ++_ranks[px];
            } else {
                _parents[px] = py;
                ++_ranks[py];
            }

            return true;
        }
    }
}
